ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp

```diff
@@ -12,17 +12,73 @@
 #include "VnodeUtilities.hpp"
 
 
+struct VirtualizationRoot
+{
+    bool                        inUse;
+    // If this is a nullptr, there is no active provider for this virtualization root (offline root)
+    PrjFSProviderUserClient*    providerUserClient;
+    int                         providerPid;
+    // For an active root, this is retained (vnode_get), for an offline one, it is not, so it may be stale (check the vid)
+    vnode_t                     rootVNode;
+    uint32_t                    rootVNodeVid;
+    
+    // Mount point ID + persistent, on-disk ID for the root directory, so we can
+    // identify it if the vnode of an offline root gets recycled.
+    fsid_t                      rootFsid;
+    uint64_t                    rootInode;
+    
+    // TODO(Mac): this should eventually be entirely diagnostic and not used for decisions
+    char                        path[PrjFSMaxPath];
+};
+
 static RWLock s_rwLock = {};
 
-// Arbitrary choice, but prevents user space attacker from causing
-// allocation of too much wired kernel memory.
-static const size_t MaxVirtualizationRoots = 128;
+// Current length of the s_virtualizationRoots array
+static uint16_t s_maxVirtualizationRoots = 0;
+static VirtualizationRoot* s_virtualizationRoots = nullptr;
+
+static VirtualizationRootHandle FindRootAtVnode_Locked(vnode_t vnode, uint32_t vid, VnodeFsidInode fileId);
+static VirtualizationRootHandle FindUnusedIndex_Locked();
+static VirtualizationRootHandle InsertVirtualizationRoot_Locked(PrjFSProviderUserClient* userClient, pid_t clientPID, vnode_t vnode, uint32_t vid, VnodeFsidInode persistentIds, const char* path);
+
+bool VirtualizationRoot_IsOnline(VirtualizationRootHandle rootIndex)
+{
+    if (rootIndex < 0)
+    {
+        return false;
+    }
+    
+    bool result;
+    RWLock_AcquireShared(s_rwLock);
+    {
+        result =
+            (rootIndex < s_maxVirtualizationRoots)
+            && (nullptr != s_virtualizationRoots[rootIndex].providerUserClient);
+    }
+    RWLock_ReleaseShared(s_rwLock);
+    
+    return result;
+}
 
-static VirtualizationRoot s_virtualizationRoots[MaxVirtualizationRoots] = {};
+bool VirtualizationRoot_PIDMatchesProvider(VirtualizationRootHandle rootIndex, pid_t pid)
+{
+    bool result;
+    RWLock_AcquireShared(s_rwLock);
+    {
+        result =
+            (rootIndex >= 0 && rootIndex < s_maxVirtualizationRoots)
```

Comment:
Style nit: I see 3 different parentheses styles in this one condition, and none of them are required :-). I personally favor: only use parens when they change the order of operations, otherwise they just add visual and mental clutter.